src/pycapsule.rs

```diff
@@ -0,0 +1,417 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+use crate::Python;
+use crate::{ffi, AsPyPointer, PyAny};
+use crate::{pyobject_native_type_core, PyErr, PyResult};
+use std::ffi::{c_void, CStr};
+use std::os::raw::c_int;
+
+/// Represents a Python Capsule
+/// as described in [Capsules](https://docs.python.org/3/c-api/capsule.html#capsules):
+/// > This subtype of PyObject represents an opaque value, useful for C extension
+/// > modules who need to pass an opaque value (as a void* pointer) through Python
+/// > code to other C code. It is often used to make a C function pointer defined
+/// > in one module available to other modules, so the regular import mechanism can
+/// > be used to access C APIs defined in dynamically loaded modules.
+///
+///
+/// # Example
+/// ```
+///  use std::ffi::CString;
+///  use pyo3::{prelude::*, pycapsule::PyCapsule};
+///
+///  #[repr(C)]
+///  struct Foo {
+///      pub val: u32,
+///  }
+///
+///  let r = Python::with_gil(|py| -> PyResult<()> {
+///      let foo = Foo { val: 123 };
+///      let name = CString::new("builtins.capsule").unwrap();
+///
+///      let capsule = PyCapsule::new(py, foo, name.as_ref())?;
+///
+///      let module = PyModule::import(py, "builtins")?;
+///      module.add("capsule", capsule)?;
+///
+///      let cap: &Foo = unsafe { PyCapsule::import(py, name.as_ref())? };
+///      assert_eq!(cap.val, 123);
+///      Ok(())
+///  });
+///  assert!(r.is_ok());
+/// ```
+#[repr(transparent)]
+pub struct PyCapsule(PyAny);
+
+pyobject_native_type_core!(PyCapsule, ffi::PyCapsule_Type, #checkfunction=ffi::PyCapsule_CheckExact);
+
+impl PyCapsule {
+    /// Constructs a new capsule whose contents are `value`, associated with `name`.
+    /// `name` is the identifier for the capsule; if it is stored as an attribute of a module,
+    /// the name should be in the format `modulename.attribute`.
+    ///
```

Comment:
add something like:

It is checked at compile time that the type `T` is not zero-sized.
Rust function items need to cast to a function pointer (`fn(args) -> result`)
to be put into a capsule.